h/static/scripts/controllers/form-controller.js

```diff
@@ -0,0 +1,204 @@
+'use strict';
+
+var Controller = require('../base/controller');
+var { findRefs, setElementState } = require('../util/dom');
+var submitForm = require('../util/submit-form');
+
+/**
+ * A controller which adds inline editing functionality to forms
+ */
+class FormController extends Controller {
+  constructor(element, options) {
+    super(element, options);
+
+    setElementState(this.refs.cancelBtn, {hidden: false});
+    this.refs.cancelBtn.addEventListener('click', event => {
+      event.preventDefault();
+      this.cancel();
+    });
+
+    // List of groups of controls that constitute each form field
+    this._fields = Array.from(element.querySelectorAll('.js-form-input'))
+      .map(el => {
+        var parts = findRefs(el);
+        return {container: el, input: parts.formInput};
+      });
+
+    this.on('focus', event => {
+      var field = this._fields.find(field => field.input === event.target);
```

Comment:
> You really don't like maps, do you? 😉

I think it is more that arrays are very convenient to work with in JS for a variety of reasons, so it becomes a bit of a habit :)

> Alternatively, and even more simply, you could just not maintain this data structure at all,
> and instead just find the parent .js-form-input field when you receive a focus event:

True. There is actually now a standardized [Element#closest](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest) DOM function for this.